Redis Programmability Documentation (#1725)
redis/redis-doc: topics/functions-intro.md

# Introduction to Redis Functions
 
As of Redis 7.0, functions offer a new programmability approach.
Functions have evolved from the [Eval Scripts](/topics/eval-intro) feature that was, in turn, added in version 2.6.

You can skip the following prologue if you're new to Redis or want to jump right into the action.

## Prologue (or what's wrong with Eval Scripts?)

Prior versions of Redis made scripting available only via the `EVAL` command, which allows a Lua script to be sent for execution by the server.
One of the core use cases for [Eval Scripts](/topics/eval-intro) is implementing and exposing a richer API composed of core Redis commands and workflow logic.
Such APIs can perform conditional updates across multiple keys, possibly combining several different data types, to maintain a consistent view of entities (e.g., user profiles) somewhat like schemas in traditional databases.

> Using `EVAL` requires that the application sends the entire script for execution every time.

Reviewer: maybe one of these two lines can be elaborated a bit (before going into the next lines that lists the limitations).
like stressing that Eval scripts are part of the client application (being executed on the server), and NOT an API enhancement. which is why they're not persisted or named.

To address these needs while avoiding breaking changes to already-established and well-liked ephemeral scripts, Redis v7.0 introduces Redis Functions.

## What are Redis Functions?

Redis functions are an evolutionary step from ephemeral scripting.

Functions provide the same core functionality as scripts but are first-class software artifacts citizens of the database.
Redis manages functions as an integral part of the database and ensures their availability via data persistence and replication.
Because functions are part of the database and therefore declared before use, applications aren't required to load them during runtime nor risk aborted transactions.
An application that uses functions depends only on their APIs rather than on the embedded script logic in the database.

Whereas ephemeral scripts are considered a part of the application's domain, functions extend the database server itself with user-provided logic.
Every function has a unique user-provided name, making it much easier to call and trace its execution.

> The design of Redis Functions also attempts to demarcate between the programming language used for writing functions and their management by the server.

Reviewer: i'm still missing something in this intro that stresses that functions are here as an alternative to a schema, so if you have several keys in which the data should always be consistent, and they're always updated together, the application is expected expected to use functions to update the data, and the functions will present logical API for the data (which means the application is not expected to issue any redis native commands to access it, only the high level logical function calls).

and maybe some paragraph that compares the to modules, developed once, loaded at startup, and used repeatedly by various applications / clients.

Author: It is explained on the function-into page.

Reviewer: this is the functions-into page.

anyway, i think the key differences between the two scripts mechanisms, and what's each one should be used for, needs to be specified in both the programability into page, and each of the specific pages (to make users re-consider going to the other one).
maybe the level of detail should be different (less detail in the programability page, and more details in the functions page)

However, the choice of Lua as a language still presents many Redis users with an obstacle. 

The Redis Functions feature makes no assumptions regarding the implementation's language.
An execution engine that is part of the definition of the function handles running it.
An engine can theoretically execute functions in any language as long as it respects several rules (such as the ability to terminate an inflight function's execution).

Presently, as noted above, Redis ships with a single embedded [Lua 5.1](/topics/lua-api) engine.
There are plans to support additional engines in the future.
Redis functions can use all of Lua's available capabilities to ephemeral scripts,
with the only exception being the [Redis Lua scripts debugger](/topics/ldb).

Functions also simplify development by enabling code sharing.
Every function belongs to a single library, and any given library can consist of multiple functions.
The library's contents are immutable, and selective updates of its functions aren't allowed.
Instead, libraries are updated as a whole with all of their functions together in one operation.

> This allows calling functions from other functions within the same library.

Reviewer: ```suggestion
This allows calling functions from other functions within the same library, or sharing code between functions by using a common code in library-internal methods, that can also take language native arguments.
```

> Functions are intended to support better the use case of maintaining a consistent view for data entities through a logical schema, as mentioned above.

Reviewer: ```suggestion
Functions are intended to better support the use case of maintaining a consistent view for data entities through a logical schema, as mentioned above.
```

As such, functions are stored alongside the data itself.

> Functions are also persisted to the AOF file and synchronized from master to replicas.

Reviewer: maybe these two lines should be expanded to specify that they're automatically propagated to replicas, but not yet to cluster nodes.
but also stress that propagating them to all cluster nodes is an administrative act, and not a responsibility of the client application or the client library's cluster support.

Reviewer: i suppose we need some short section about cluster
* mention how to load them into all cluster nodes
* mention that when adding node it is done automatically

and also mention `redis-cli --functoins-rdb` for the ephemeral case.

Assuming you've saved the library's implementation in the _mylib.lua_ file, you can replace it with its (optional) description with:

```bash
$ cat mylib.lua | redis-cli -x FUNCTION LOAD Lua mylib REPLACE DESCRIPTION "My application's Hash data type enhancements"
```

Once loaded, you can call the library's functions with `FCALL`:

```
redis> FCALL my_hgetall 1 myhash
1) "myfield"
2) "some value"
3) "another_field"
4) "another value"
redis> FCALL my_hlastmodified 1 myhash
"1640772721"
```

You can also get the library's details with the `FUNCTION LIST` command:

```
redis> FUNCTION LIST
1) 1) "library_name"
   2) "mylib"
   3) "engine"
   4) "LUA"
   5) "description"
   6) "My application's Hash data type enhancements"
   7) "functions"
   8) 1) 1) "name"
         2) "my_hset"
         3) "description"
         4) (nil)
      2) 1) "name"
         2) "my_hgetall"
         3) "description"
         4) (nil)
      3) 1) "name"
         2) "my_hlastmodified"
         3) "description"
         4) (nil)
```

You can see that it is easy to update our library with new capabilities.

> ## Reusing code in the library

Reviewer: we're missing a section about flags.
maybe add it in a follow up PR (so add a todo comment at the top or something).
when we do add it, keep in mind that the're soon gonna serve EVAL too.

20075:M 1 Jan 2022 16:54:01.309 # Only one key name is allowed
```

## Function Flags

Looking at the example above, the library introduces 3 functions, `my_hset`, `my_hgetall`, and `my_hlastmodified`. The last 2 function are not perforning any writes, so we should be able to run them using `FCALL_RO` on a `readonly` replica. But if we try to invoke them we will get the following error:

```
> FCALL_RO my_hgetall 1 myhash
(error) ERR Can not execute a function with write flag using fcall_ro.
```

This is happened because, by default, a function can perfrom read and writes commands, and Redis has no way to know what a function might try to do. So Redis assumes the wrost and not allow running the function on the following cases:

1. On readonly replica
2. Using `EVAL_RO`

> 3. If a disc error detacted

Reviewer: disc -> disk

```lua
redis.register_function('my_hset', my_hset)
redis.register_function{
  function_name='my_hgetall',
  callback=my_hgetall,
  flags={'no-writes'}
}
redis.register_function{
  function_name='my_hlastmodified',
  callback=my_hlastmodified,
  flags={'no-writes'}
}
```

Now Redis will allows us to run both, `my_hgetall` and `my_hlastmodified`, using `FCALL_RO` and on `readonly` replica:

```
redis> FCALL_RO my_hgetall 1 myhash
1) "myfield"
2) "some value"
3) "another_field"
4) "another value"
redis> FCALL_RO my_hlastmodified 1 myhash
"1640772721"
```

### Default behaviure of functions (if no flags is given):

1. Allow functions to read and write
2. Do not run functions on OOM
3. Do not run functions on stale replica
4. Allow functions on cluster

> ### Supported Function Flags:

Reviewer: the above explanation, and examples, and maybe even the defaults probably belong here.
but the dry explanation of the flags should go to the other page, since it's soon gonna serve EVAL too.

* `no-writes` - indicating the function perform no writes which means that it is OK to run it on:
   * read-only replica
   * Using FCALL_RO
   * If disk error detected
   

>    It will not be possible to run a function in those situations unless the function turns on the `no-writes` flag

Reviewer: it might be obvious, but let's also state that execution of write commands will be refused.
unlike the `allow-oom` flag

> * `allow-oom` - indicate that its OK to run the function even if Redis is in OOM state, if the function will not turn on this flag it will not be possible to run it if OOM reached (even if the function declares `no-writes` and even if `fcall_ro` is used). If this flag is set, any command will be allow on OOM (even those that is marked with CMD_DENYOOM). The assumption is that this flag is for advance users that knows its meaning and understand what they are doing, and Redis trust them to not increase the memory usage. (e.g. it could be an INCR or a modification on an existing key, or a DEL command)

Reviewer: * let's break this into multiple lines (too long).
* i think we need to specify at least one what OOM means (out of memory).
* `will be allow on OOM` -> `will be allowed on OOM`
* `even those that is marked with CMD_DENYOOM` -> `(even ones that are normally not allowed in that state)` 
* `the assumption....` - don't think we need to state that here

> * `allow-state` - indicate that its OK to run the function on stale replica, in this case we will also make sure the function is only perform `stale` commands and raise an error if not.

Reviewer: `we` -> `Redis`
```suggestion
* `allow-state` - indicate that its OK to run the function on a stale replica, in this case calls to Redis commands that are not allowed in this state will error when the script is executed on a stale replica.
```

> * `no-cluster` - indicate to disallow running the function if cluster is enabled.

Reviewer: `cluster is enabled` -> `cluster mode is enabled`
or `on Redis cluster`

When you register a function, the server does not know how it accesses the database.
On the other hand, the server must also enforce resource usage policies and maintain data consistency for functions.
By default, and unless instructed differently, Redis assumes that all functions read and write data.

Function flags tell the server that its assumptions are wrong.
They allow you to inform Redis that your function is an exception.
Use care when registering a function with flags, which may negatively impact if misused.

Let's see how flags work.
Our example had two functions that only read data, so we can try running them using `FCALL_RO` against a read-only replica. 

```
redis > FCALL_RO my_hgetall 1 myhash
(error) ERR Can not execute a function with write flag using fcall_ro.
```

Redis returns this error because a function can, in theory, perform both read and write operations on the database.
As a safeguard and by default, Redis assumes that the function does both, so it blocks its execution.
The server will reply with this error in the following cases:

1. Executing a function with `FCALL` against a read-only replica.
2. Using `FCALL_RO` to execute a function.

> 3. A disk error was detected.

Reviewer: maybe add a few more works besides "disk" (like refer to persistence)
